### PR TITLE
LRQA 81155

### PR DESCRIPTION
**Pull request by kyle-miho**
Comment from original PR: https://github.com/kyle-miho/liferay-portal/pull/1018
https://liferay.atlassian.net/browse/LRQA-81155

Reworked this to reuse testFactoryConfiguration as you requested. Just needed to have a variable count for the latch depending on how many configs there were. Seems they are all updated when there is an existing factory config since updated is called that many times. So since it was hardcoded to 1, it really wasn't waiting at all for the new config to be updated. 

The only thing I couldn't figure out is in the update test. Still seems like a need a 1 second buffer when updating an existing config or else the test gets stuck. The count still gets set to 2 though not sure what's going on there. So I still have a one second sleep in just that test. The multi configs test has no issues.

### CODE REVIEW

**Comment by kyle-miho** [comment]
Theres still a pause in, but its probably due to the latch being used incorrectly for FactoryConfiguration, could someone help look into it?

**Comment by Kvale1733** [comment]
ci:test:relevant

File: modules/apps/static/portal-file-install/portal-file-install-test/src/testIntegration/java/com/liferay/portal/file/install/deploy/test/FileInstallConfigTest.java [hunk 1]
```diff
@@ -414,10 +452,24 @@ private void _deleteConfiguration() throws Exception {
...
 		String factoryConfigurationName = StringBundler.concat(
 			StringUtil.randomId(), CharPool.DASH, StringUtil.randomId());
 
+		String testValue = StringUtil.randomId();
+
+		_testFactoryConfiguration(
+			separator, factoryConfigurationName, testValue);
+
+		return _configuration;
+	}
```
Comment (yuhai) [inline modules/apps/static/portal-file-install/portal-file-install-test/src/testIntegration/java/com/liferay/portal/file/install/deploy/test/FileInstallConfigTest.java:467]: @kyle-miho, for the kind of method _test*(), please don't return any object and keep the original `void`. Like other methods `private Configuration _createConfiguration()`, we may return Object when using the kind of methods.

**Comment by yuhai** [comment]
Hi @kyle-miho, please the above comments. 

I want to firstly answer you mention the [issue](https://github.com/liferay-core-infra/liferay-portal/pull/6079#issuecomment-1802978533). 

The fix added `Thread.sleep(1000);` to fix the thread waiting issue when run `testMultipleFactoryConfigurations()`. The happened reason as the followings:
After create the original config file(for example, test~A.config), `Scanner#updateChecksum()` didn't been executed. After change the original config file content("testValue2"), `Scanner#updateChecksum()` was executed and it used the updated config file to calculate `storedChecksum`. When the monitor thread `fileinstall-directory-watcher` run, it used the new updated config file(same config file content with storedChecksum used) to calculate newChecksum. The newChecksum will equal to storedChecksum, please see `Scanner#_processFiles()` logic. It will return empty files so that missing once update and the waiting happened.

So adding `Thread.sleep(1000);` is reasonable to me.

Back to the fix, please see the below suggestions:
1. Remove the commit(5f359b19efb1f6c79518a13046181937e7009d20), because the following `dictionary.get()` will break the test if dictionary is null.
2. We may move `_testFactoryConfiguration(char separator)` method for two methods, one is `private Configuration _createFacotryConfiguration()`, another is `private void _assertFacotryConfiguration()`. And then we may invoke the two methods in the test methods. For test `DeployFactoryConfigs`, we may get the two configurations and configurationPaths, and then delete them in the test methods. Just like:
```
public void testMultipleFactoryConfigurations() throws Exception {
       Configuration configuration1;
       Path configurationPath1;
Configuration configuration2;
       Path configurationPath2;
      
try {
configuration1 = _createFacotryConfiguration();
configurationPath1 = _configurationPath;
      _assertFacotryConfiguration(...);

configuration2 = ;
configurationPath2 =;
_assertFacotryConfiguration(...);
}
finally{
     delete...
}

}
```
3. Remove the count related change from `_createFactoryConfiguration()`, we may create one method parm named int count to `private Configuration _createFacotryConfiguration(.., int count)` 
4. (Optional) Rename `testMultipleFactoryConfigurations()` for `testTwoFactoryConfigurations()`

I am going to close the pr. Please kindly let me know if this is not clear to you.

Thanks,
Hai

**Pull request closed without merging**